From db7715407efa43cd2a66caed67c02d8f7bb90b35 Mon Sep 17 00:00:00 2001
From: Paul Wouters <pwouters@redhat.com>
Date: Tue, 12 May 2020 12:56:38 -0400
Subject: [PATCH] nss: Set NSS_PKCS11_2_0_COMPAT to ensure using the compat
 interface for now.

This might resolve the issue seen in https://github.com/libreswan/libreswan/issues/334

As per conversation with Bob:

	The issue is building with nss 3.52, If you build with 3.51 and run with
	3.52 you won't run into the issue.  It's the default for the definition
	of CK_GCM_PARAMS. The spec and the released headers were different from
	OASIS. In that case, the header is authoritative and we used the spec NSS
	needs to move the new definition, but doing so will break things that
	compile with NSS. To get around it you can add -DNSS_PKCS11_2_0_COMPAT
	or include it in your .c file

	Long term, you'll actually want to move the the AEAD interface.
	There's a new PKCS #11 interface that allows you to operate multiple AEAD
	operations on a single key. It allows token IV generation. I added new
	wrappers in 3.52 to handle the differences between tokens and mechanism
---
 lib/libswan/ike_alg_encrypt_nss_gcm_ops.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/lib/libswan/ike_alg_encrypt_nss_gcm_ops.c b/lib/libswan/ike_alg_encrypt_nss_gcm_ops.c
index 93a027089a..571913cc1e 100644
--- a/lib/libswan/ike_alg_encrypt_nss_gcm_ops.c
+++ b/lib/libswan/ike_alg_encrypt_nss_gcm_ops.c
@@ -16,6 +16,12 @@
 #include <stdio.h>
 #include <stdlib.h>
 
+/*
+ * Special advise from Bob Relyea - needs to go before any nss include
+ *
+ */
+#define NSS_PKCS11_2_0_COMPAT 1
+
 #include "lswlog.h"
 #include "lswnss.h"
 #include "prmem.h"
